XD-3482 Add jdbc sink

trisberg opened the pull request:
- based on Eric's PR #40
- change to use JdbcMessageHandler from Spring Integration

ericbottard committed:
XD-3482: Add jdbc sink

trisberg committed:
XD-3482: Add jdbc sink

- change to use JdbcMessageHandler from Spring Integration

ericbottard commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,179 @@
+		}
+		JdbcMessageHandler jdbcMessageHandler = new JdbcMessageHandler(dataSource,
+				generateSql(properties.getTableName(),columnExpressionVariations.keySet()));
+		jdbcMessageHandler.setSqlParameterSourceFactory(
```
Wondering if/how `ExpressionEvaluatingSqlParameterSourceFactory` could be used here.

trisberg replied:
Don't think it supports multiple expressions for a column

ericbottard replied:
Could we fake it by constructing a single expression using the elvis operator?

trisberg replied:
Added some code to handle JsonPropertyAccessor.ToStringFriendlyJsonNode so I think we are stuck with our own implementation for now.

ericbottard commented:
Why drop the batching support? Could have our own, at least till SI provides what we need, WDYT?

trisberg commented:
I thought we decided to wait for it to be in SI -- https://jira.spring.io/browse/INT-3530

trisberg committed:
XD-3482: Add jdbc sink

- add logging for missing values for a column

trisberg committed:
XD-3482: Add jdbc sink

- add support for handling JsonPropertyAccessor.ToStringFriendlyJsonNode values

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/DefaultInitializationScriptResource.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Eric Bottard
+ */
+public class DefaultInitializationScriptResource extends ByteArrayResource {
```
Any arguments that this class should be `public`, but not an inner one in the `JdbcSinkConfiguration` ?

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/DefaultInitializationScriptResource.java:
```diff
@@ -0,0 +1,56 @@
+			if (i++ > 0) {
+				result.append(", ");
+			}
+			result.append(column).append(" VARCHAR(2000)");
```
Does `VARCHAR` work for any RDBMS?
And why `2000`, but not `4000`?..

ericbottard replied:
This will be enhanced in a subsequent PR. This mimics the current XD behavior

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkApplication.java:
```diff
@@ -0,0 +1,18 @@
+@SpringBootApplication
+public class JdbcSinkApplication {
+
+	public static void main(String[] args) throws InterruptedException {
```
The `throws` looks like overhead here.

trisberg replied:
More like noise, but doesn't seem necessary. Removing it.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+	public static final Object NOT_SET = new Object();
+
+	@Autowired(required = false)
+	private SpelExpressionParser spelExpressionParser = new SpelExpressionParser();
```
Interesting construction...
So, `@Autowired` will overwrite this `new` on the injection phase ?

ericbottard replied:
Yeah, I don't think we need the autowire here. This was added when I thought we needed SI bits, but those are actually carried by the `EvaluationContext`, not the parser itself.

trisberg replied:
Removing the autowire.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkApplication.java:
```diff
@@ -0,0 +1,18 @@
+package org.springframework.cloud.stream.module.jdbc;
```
No `Copyright`.
Is that intentional?

trisberg replied:
Adding it.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	@ServiceActivator(autoStartup = "false", inputChannel = Sink.INPUT)
+	public void handleMessage(Message<?> msg) {
```
This isn't only redundant, but also some overhead.
The `@ServiceActivator` declaration should be moved to the `jdbcMessageHandler` `@Bean` definition.
The same concern which I raised in the Cassandra Sink PR before.

trisberg replied:
Ahh, yes, moving it.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+		dataSourceInitializer.setDatabasePopulator(databasePopulator);
+		if ("true".equals(properties.getInitialize())) {
+			databasePopulator.addScript(new DefaultInitializationScriptResource(properties));
+		} else {
```
Minor, of course, but anyway `else` on new line.
Some IDE code formatting settings vulnerabilities.

ericbottard replied:
+1

trisberg replied:
Fixed.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+		databasePopulator.setIgnoreFailedDrops(true);
+		dataSourceInitializer.setDatabasePopulator(databasePopulator);
+		if ("true".equals(properties.getInitialize())) {
+			databasePopulator.addScript(new DefaultInitializationScriptResource(properties));
```
Looking at this one more time, I really don't like this `DefaultInitializationScriptResource`.
You should just build an appropriate String from the properties and use its `getBytes()` for the regular `ByteArrayResource`.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	@Bean
+	public JdbcOperations jdbcOperations(DataSource dataSource) {
```
This is redundant bean

ericbottard replied:
You mean it's already added by boot?

artembilan replied:
No, it isn't. I just mean that we don't use it at all.
Well, I see you use it from test-case. 
Do, we really need this extra (and very simple) bean in production anyway ? :smile:

ericbottard replied:
Ah. Maybe it's a leftover from my previous implementation then. I'll have a look

trisberg replied:
Removing it.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+							for (Expression spel : spels) {
+								try {
+									value = spel.getValue(evaluationContext, message);
+								}
```
Looks like you have missed `break` here if the first expression evaluates correctly.

trisberg replied:
Adding a break.

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+
+	@PostConstruct
+	public void afterPropertiesSet() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(beanFactory);
```
Should be changed to the `ExpressionUtils.createStandardEvaluationContext(beanFactory)` - creates `StandardEvaluationContext` if there is no one in the ctx.

ericbottard replied:
Then what's the point of `IntegrationContextUtils.getEvaluationContext`?
We want the one from SI, where we know some useful `PropertyAccessors` will have been registered

artembilan replied:
The `IntegrationContextUtils` deals only with the `BeanFactory`.
The `ExpressionUtils` delegates to the `IntegrationContextUtils` for the first attempt and fallbacks to the `StandardEvaluationContext` if there is no that one in the `BF`.
Just take a look to its source code.

I agree that it feels weird to have several similar important utility methods, but that is the historical result. We can revise it in the future release, but I'm not sure that it will simplify our life, just because we will be forced to refactor everything one more time :smile:

trisberg replied:
Changed to use `ExpressionUtils`

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+	public static class Nested {
+		@Bean
+		@ConfigurationPropertiesBinding
+		public ShorthandMapConverter shorthandMapConverter() {
```
I wonder how is it resolved if we don't have here `@Configuration` or even `@Component` for `light` `@Bean`?..

ericbottard replied:
You mean on the class?

artembilan replied:
Yes, `@Configuration` on this `Nested`  class.
Just explain for me how it works without that.
Thanks

ericbottard replied:
Nested classes of `@Configuration` classes are scanned automatically (I discovered that myself recently). Even more, inner classes should not be annotated with `@Configuration` if it is profile driven (I discovered that when looking at https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-admin/src/main/java/org/springframework/cloud/dataflow/admin/config/CloudConfiguration.java#L45-L69

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/ShorthandMapConverter.java:
```diff
@@ -0,0 +1,66 @@
+			}
+			// Split on colon, if not preceded by backslash
+			String[] keyValuePair = unescaped.split("(?<!\\\\):");
+			Assert.isTrue(keyValuePair.length <= 2, "'" + unescaped + "' could not be parsed to a 'key:value' pair or simple 'key' with implicit value");
```
Very long line of code.
Minor,  of course, but I wonder if we follow with some code style rules...

ericbottard replied:
will split

trisberg replied:
done

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/ShorthandMapConverter.java:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @author Eric Bottard
+ */
+public class ShorthandMapConverter implements Converter<String, Map<String, String>>, ConditionalConverter{
```
Seems for me this is a good candidate to be moved somewhere into `common`

artembilan replied:
Pay attention, please #56

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/SupportsShorthands.java:
```diff
@@ -0,0 +1,34 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.FIELD, ElementType.METHOD})
+public @interface SupportsShorthands {
```
Why can't we bypass this one?

ericbottard replied:
What do you mean?

artembilan replied:
I mean if we would be able to live just with the `Converter<String, Map<String, String>>` condition, without an extra annotation do not forget to add to accept the `ShorthandMapConverter` for the property.
Or do we have here the collision with `StringToPropertiesConverter`?

ericbottard replied:
Not only that, but not all `Properties` injection points want the "shorthand" behavior that `ShorthandMapConverter` comes with

artembilan replied:
OK. That's fine with me. 
So, shouldn't we move these both guys somewhere to `common` to be reusable, like that Groovy Script variables case ?

ericbottard replied:
We could indeed. I would wait for another consumer of this specific need though, what do you think?

artembilan commented on jdbc-sink/src/test/java/org/springframework/cloud/stream/module/jdbc/ShorthandMapConverterTest.java:
```diff
@@ -0,0 +1,79 @@
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.collection.IsMapContaining.*;
+import static org.junit.Assert.*;
```
Is it OK to use asterisk imports?

ericbottard replied:
This is weird. Unless I am mistaken, I am using the sanctioned `spring-framework.xml` IDEA settings, which have static star imports set at 1:
![image](https://cloud.githubusercontent.com/assets/313494/10817637/781f327c-7e3b-11e5-9c69-316e54b6d660.png)

Or is this IDEA playing tricks on me again?

artembilan replied:
I think you are correct here: https://github.com/spring-projects/spring-framework/blob/master/src/idea/spring-framework.xml#L16.

And SF team really uses asterisk imports: 
https://github.com/spring-projects/spring-framework/blob/master/spring-websocket/src/test/java/org/springframework/web/socket/client/WebSocketConnectionManagerTests.java#L37

https://github.com/spring-projects/spring-framework/blob/master/spring-jms/src/test/java/org/springframework/jms/annotation/EnableJmsTests.java#L47

But should we do the same in our projects?..

artembilan commented on jdbc-sink/src/test/java/org/springframework/cloud/stream/module/jdbc/ShorthandMapConverterTest.java:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @author Eric Bottard
+ */
+public class ShorthandMapConverterTest {
```
Must be in plural. Isn't it ?

ericbottard replied:
correct

artembilan commented on jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSinkConfiguration.java:
```diff
@@ -0,0 +1,217 @@
+		for (Map.Entry<String, String> entry : properties.getColumns().entrySet()) {
+			String value = entry.getValue();
+			columnExpressionVariations.add(entry.getKey(), spelExpressionParser.parseExpression(value));
+			if (!value.startsWith("payload")) {
```
I find this like a redundant trick.
Let's consider if I'd like to use some `header` from message or even just `new java.util.Date()`.
Right, I know that you don't (shouldn't) reach the second expression if the first one evaluates fine, but with this shorthand "hell" we can't reuse an existing `ExpressionEvaluatingSqlParameterSourceFactory`.

For the `payload` prefix I can agree for the case like `columns=a,b`

ericbottard replied:
Not sure what you mean. So, what do you suggest?

artembilan replied:
I suggest for case of `payload` always request this prefix:

```
columns=a: payload.a, b
```

where `b` will be converted to the `b: payload.b`, like you do right now.
Only single `expression` and, therefore, `ExpressionEvaluatingSqlParameterSourceFactory` gain.
Just because I will be able to do this:

```
columns=a: payload.a, b: headers.b
```

ericbottard replied:
The whole point of the "shorthand" trick is to quickly reference a set of props to column mappings.
If we don't implicitly assume `payload.` as a prefix, then there is no point in the shorthand notation, as `payload.x` is certainly what would be used in 95% of the cases.

I suggest to merge as is, and we can revisit, for example maybe by allowing `ExpressionEvaluatingSqlParameterSourceFactory` to take several alternatives

trisberg committed:
XD-3482: Add jdbc sink

- addressing review comments

- adding tests for variety of payload types like Tuple, Map, JSON string

trisberg committed:
XD-3482: Add jdbc sink

- fix JSON string test

artembilan commented:
> I thought we decided to wait for it to be in SI

Guys, what would you like to see there?
Any thought how to address? 

Thanks

ericbottard commented:
Merged as 9643875 alongside https://github.com/spring-cloud/spring-cloud-dataflow/commit/16c9db1a2138ca597c8a8dbd13966e3cece26c98
